### PR TITLE
add option to enable or disable from default metrics

### PR DESCRIPTION
**Pull request by aanm**
This PR adds metric levels which allows for users to specify which metrics they want to choose. Right now the bpf syscalls metrics will be the only ones not considered the default.

```release-note
Add option to enable or disable from default metrics (`+metric_foo` to enable `metric_foo`, `-metric_bar` to disable `metric_bar`)
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7896)

### CODE REVIEW

File: daemon/daemon_main.go [hunk 1]
```diff
@@ -602,6 +602,9 @@ func init() {
 	flags.MarkHidden(option.MaxCtrlIntervalName)
 	option.BindEnv(option.MaxCtrlIntervalName)
 
+	flags.Int(option.MetricsLevel, option.MetricLevelCore, "Level of metrics that Cilium exposes.")
```
Comment (tgraf) [inline daemon/daemon_main.go]: What about we  do a StringSlice and allow to define a mask of what should be recorded?

`--metrics=kvstore-histogram --metrics=bpf-syscall [...]`

Comment (ianvernon) [reply daemon/daemon_main.go]: Agreed, this is more user-friendly, as the choices for integers are arbitrary.

File: pkg/option/config.go [hunk 2]
```diff
@@ -442,6 +442,22 @@ const (
...
+	// MetricsLevel represents the level of metrics that Cilium should expose
+	// the higher the level more metrics it exposes.
+	MetricsLevel = "metrics-level"
+)
+
+// Metric Levels
+const (
+	// MetricLevelNone should not expose any metric!
+	MetricLevelNone int = iota
+	// MetricLevelCore should expose core TBD
```
Comment (joestringer) [inline pkg/option/config.go]: core TBD?

Comment (aanm) [reply pkg/option/config.go]: core, To Be Decided :)

**Comment by aanm** [comment]
test-me-please

File: daemon/daemon_main.go [hunk 3]
```diff
@@ -602,6 +602,9 @@ func init() {
 	flags.MarkHidden(option.MaxCtrlIntervalName)
 	option.BindEnv(option.MaxCtrlIntervalName)
 
+	flags.StringSlice(option.SubsysMetrics, metrics.DefaultSubsystems(), "Subsystems metrics that Cilium exposes.")
```
Comment (ianvernon) [inline daemon/daemon_main.go]: remove the period

**Comment by aanm** [comment]
test-me-please

File: pkg/bpf/bpf_linux.go [hunk 4]
```diff
@@ -128,14 +139,19 @@ func LookupElement(fd int, key, value unsafe.Pointer) error {
 		value: uint64(uintptr(value)),
 	}
 
-	duration := spanstat.Start()
+	var duration *spanstat.SpanStat
+	if option.Config.IsSubsysMetricEnabled(metrics.SubsystemBPFMask) {
```
Comment (ianvernon) [inline pkg/bpf/bpf_linux.go]: This isn't blocking, but I feel like this check would be way simpler overall if the value being checked was a variable set at the package level? We can use `option.Config` to gather all configuration from daemon bootstrap, then at bootstrap of this package, take in the value from `option.Config` and set it as a package-scoped variable.

**Review by ianvernon — APPROVED** [review]
I have no blocking comments - it makes sense to get this merged sooner rather than later so we can get it into `v1.5`. I do think that it would be better for code readability to set the metrics configuration as a package-level variable from `pkg/option` instead of having to call back into `option.Config` every time we need to do a lookup.

File: pkg/bpf/bpf_linux.go (hunk 4 above)
Comment (aanm) [reply pkg/bpf/bpf_linux.go]: @ianvernon That's going definitely going to be a follow up as the performance improves as well.

```
5506 ns/op vs 5365 ns/op
```

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -55,6 +56,98 @@ const (
...
+	// SubsystemNodesMask is the mask used for the node manager
+	SubsystemNodesMask
+
+	// SubsystemTriggersMask is the mask used in the pkg/trigger
+	SubsystemTriggersMask
+)
+
+// DefaultSubsystems returns the default subsystems that will be used to expose
+// metrics.
+func DefaultSubsystems() []string {
```
Comment (tgraf) [inline pkg/metrics/metrics.go]: I don't think the granularity of entire subsystems will be sufficient. Most subsystems will be scalable and less scalable metrics and users are likely to opt out of less scalable metrics.

I see two paths that make sense:

1. High-level flag that disables less scalable metrics such as histograms  
2. A per metric systems where we have a default `[]string{metric1,metric2,metric3}` and then allow to modify that with `--metrics=+metric2 --metrics=-metric3` to allow enabling or disabling individual metrics without requiring to be aware of the default all the time.

File: pkg/option/config.go [hunk 6]
```diff
@@ -965,6 +973,11 @@ func (c *DaemonConfig) IsFlannelMasterDeviceSet() bool {
 	return len(c.FlannelMasterDevice) != 0
 }
 
+// IsSubsysMetricEnabled returns true if the given subsystem metric is enabled.
+func (c *DaemonConfig) IsSubsysMetricEnabled(subsystem uint64) bool {
```
Comment (tgraf) [inline pkg/option/config.go]: What's the benefit of a uint64 mask over a struct with booleans? It seems simpler to be able to:
```
if option.Config.EnabledMetrics.BpfSyscalls {
```

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/metrics/metrics.go (hunk 5 above)
Comment (aanm) [reply pkg/metrics/metrics.go]: @tgraf I hope that I'm not understanding 2) correctly, we will give the users to select over 50 different metrics, this sounds extremely overly complicated. If users don't want those metrics from being considered in prometheus they can configure prometheus to do that.

Comment (tgraf) [reply pkg/metrics/metrics.go]: How does your current approach solve the problem of providing metrics while being able to disable the ones that are not scalable? I don't think we want users to get too involved. We want to have a good set of defaults and allow developers and advanced to enable individual metrics. We can also to `+grp-XXX` to enable individual groups.

I think subsystem is an arbitrary grouping that will be incredibly limiting if users want to find a good trade-off between visibility and load.

Comment (aanm) [reply pkg/metrics/metrics.go]: >How does your current approach solve the problem of providing metrics while being able to disable the ones that are not scalable?

Scalable where? In prometheus or Cilium? In Cilium the only non scalable metrics are the bpf related ones. In prometheus, users can specify which metrics they want to disable in its configuration.

File: pkg/metrics/metrics.go [hunk 7]
```diff
@@ -557,72 +365,594 @@ var (
...
 
-	MustRegister(KVStoreOperationsDuration)
-	MustRegister(KVStoreEventsQueueDuration)
+	return c, collectors
+}
 
-	MustRegister(FQDNGarbageCollectorCleanedTotal)
+func init() {
+	MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: Namespace}))
+	// TODO: Figure out how to put this into a Namespace
```
Comment (joestringer) [inline pkg/metrics/metrics.go:958]: What's the story behind this?

File: pkg/metrics/interfaces.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+func (ov *observerVec) WithLabelValues(...string) prometheus.Observer { return NoOpObserver }
+
+func (ov *observerVec) CurryWith(prometheus.Labels) (prometheus.ObserverVec, error) {
+	return NoOpObserverVec, nil
+}
+func (ov *observerVec) MustCurryWith(prometheus.Labels) prometheus.ObserverVec {
+	return NoOpObserverVec
+}
+
+// Gauge
```
Comment (joestringer) [inline pkg/metrics/interfaces.go:113]:
```suggestion
// gauge is a prometheus metric that acts as a no-op for use when metrics for subsystems are disabled.
```

**Review by joestringer — APPROVED** [review]
BPF bits LGTM, One todo I noticed below and a minor nit.

**Comment by joestringer** [comment]
Travis failure is legit build failure.

**Review by tgraf — APPROVED** [review]
Awesome work!

File: pkg/metrics/metrics.go (hunk 7 above)
Comment (aanm) [reply pkg/metrics/metrics.go:958]: It was already there. git blames @raybejjani on this one

File: pkg/metrics/interfaces.go (hunk 8 above)
Comment (aanm) [reply pkg/metrics/interfaces.go:113]: @joestringer I've added that to make it easier the separation of all structures. I believe they are all self-explanatory. I've added the following warning to the one that is not.

```golang
// *WARNING*: Desc returns nil so do not register this metric into prometheus
// default register.
func (m *metric) Desc() *prometheus.Desc  { return nil }
```

**Review by ianvernon — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
Known flake #7921

**Pull request merged**